infoschema/builder.go

```diff
@@ -89,21 +87,37 @@ func (b *Builder) ApplyDiff(m *meta.Meta, diff *model.SchemaDiff) ([]int64, erro
 				)
 			}
 			oldDBInfo := b.copySchemaTables(oldRoDBInfo.Name.L)
-			b.applyDropTable(oldDBInfo, oldTableID)
+			tmpIDs = b.applyDropTable(oldDBInfo, oldTableID, tmpIDs)
 		} else {
-			b.applyDropTable(dbInfo, oldTableID)
+			tmpIDs = b.applyDropTable(dbInfo, oldTableID, tmpIDs)
+		}
+
+		if oldTableID != newTableID {
+			// Update tblIDs only when oldTableID != newTableID because applyCreateTable() also updates tblIDs.
+			tblIDs = tmpIDs
 		}
 	}
 	if tableIDIsValid(newTableID) {
 		// All types except DropTableOrView.
-		err := b.applyCreateTable(m, dbInfo, newTableID, allocs, diff.Type)
+		var err error
+		tblIDs, err = b.applyCreateTable(m, dbInfo, newTableID, allocs, diff.Type, tblIDs)
 		if err != nil {
 			return nil, errors.Trace(err)
 		}
 	}
 	return tblIDs, nil
 }
 
+func appendAffectedIDs(affected []int64, tblInfo *model.TableInfo) []int64 {
+	if pi := tblInfo.GetPartitionInfo(); pi != nil {
+		for _, def := range pi.Definitions {
+			affected = append(affected, def.ID)
+		}
+	}
+	affected = append(affected, tblInfo.ID)
```

Comment:
Could we put this line to line111? That feels like it can be checked earlier if the DML is related to recent mode changes